File: backend/src/main/java/com/technicalescaperoom/backend/service/admin/TeamExcelImportService.java
package com.technicalescaperoom.backend.service.admin;

import com.technicalescaperoom.backend.dto.admin.*;
import com.technicalescaperoom.backend.entity.Event;
import com.technicalescaperoom.backend.entity.Player;
import com.technicalescaperoom.backend.entity.Team;
import com.technicalescaperoom.backend.enums.EventStatus;
import com.technicalescaperoom.backend.enums.PlayerStatus;
import com.technicalescaperoom.backend.enums.TeamStatus;
import com.technicalescaperoom.backend.exception.ResourceNotFoundException;
import com.technicalescaperoom.backend.repository.EventRepository;
import com.technicalescaperoom.backend.repository.PlayerRepository;
import com.technicalescaperoom.backend.repository.TeamRepository;
import com.technicalescaperoom.backend.service.GameStateService;
import com.technicalescaperoom.backend.util.TeamCodeGenerator;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.poi.ss.usermodel.*;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.io.InputStream;
import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class TeamExcelImportService {

    private final EventRepository eventRepository;
    private final TeamRepository teamRepository;
    private final PlayerRepository playerRepository;
    private final TeamCodeGenerator teamCodeGenerator;
    private final GameStateService gameStateService;
    private final AdminAuditService adminAuditService;

    private static final int MAX_TEAM_NAME_LENGTH = 100;
    private static final int MAX_PLAYER_NAME_LENGTH = 100;

    /**
     * Parse and validate an uploaded Excel file, returning a preview without importing.
     */
    public TeamImportPreviewDto parseAndValidate(Long eventId, MultipartFile file) {
        Event event = eventRepository.findById(eventId)
                .orElseThrow(() -> new ResourceNotFoundException("Event not found for ID " + eventId));

        validateEventEditable(event);
        validateFileFormat(file);

        List<TeamImportRow> rows = parseExcelFile(file);
        validateRows(rows, eventId);

        int validCount = (int) rows.stream().filter(TeamImportRow::isValid).count();
        int invalidCount = (int) rows.stream().filter(r -> !r.isValid()).count();
        int duplicateCount = (int) rows.stream()
                .filter(r -> r.getValidationErrors().stream().anyMatch(e -> e.contains("Duplicate")))
                .count();

        List<String> globalErrors = new ArrayList<>();
        if (rows.isEmpty()) {
            globalErrors.add("No data rows found in the Excel file.");
        }
        if (validCount == 0 && !rows.isEmpty()) {
            globalErrors.add("All rows have validation errors. No teams can be imported.");
        }

        List<String> warnings = new ArrayList<>();
        if (invalidCount > 0) {
            warnings.add(invalidCount + " row(s) have validation errors and will be skipped during import.");
        }

        return TeamImportPreviewDto.builder()
                .totalRows(rows.size())
                .validRows(validCount)
                .invalidRows(invalidCount)
                .duplicateRows(duplicateCount)
                .rows(rows)
                .errors(globalErrors)
                .warnings(warnings)
                .importReady(validCount > 0 && globalErrors.isEmpty())
                .build();
    }

    /**
     * Import validated teams from a previously parsed Excel preview.
     * Entire operation is atomic — any failure rolls back all created teams.
     */
    @Transactional
    public TeamImportResultDto importTeams(Long eventId, MultipartFile file,
                                           com.technicalescaperoom.backend.config.security.AdminPrincipal principal) {
        Event event = eventRepository.findById(eventId)
                .orElseThrow(() -> new ResourceNotFoundException("Event not found for ID " + eventId));

        validateEventEditable(event);

        List<TeamImportRow> rows = parseExcelFile(file);
        validateRows(rows, eventId);

        List<TeamImportRow> validRows = rows.stream()
                .filter(TeamImportRow::isValid)
                .collect(Collectors.toList());

        if (validRows.isEmpty()) {
            return TeamImportResultDto.builder()
                    .teamsCreated(0)
                    .playersCreated(0)
                    .duplicatesSkipped(rows.size())
                    .errorsEncountered(rows.size())
                    .errors(List.of("No valid rows to import."))
                    .summary("Import aborted: no valid team rows found.")
                    .build();
        }

        int teamsCreated = 0;
        int playersCreated = 0;
        List<String> createdCodes = new ArrayList<>();
        List<String> importErrors = new ArrayList<>();

        long currentTeamCount = teamRepository.countByEventId(eventId);

        for (TeamImportRow row : validRows) {
            try {
                // Generate unique team code
                long countForCode = currentTeamCount + teamsCreated;
                String teamCode = teamCodeGenerator.generateUniqueTeamCode(countForCode,
                        code -> teamRepository.existsByEventIdAndTeamCode(eventId, code));

                Team team = Team.builder()
                        .event(event)
                        .teamCode(teamCode)
                        .teamName(row.getTeamName().trim())
                        .status(TeamStatus.REGISTERED)
                        .build();

                Team savedTeam = teamRepository.save(team);

                Player p1 = Player.builder()
                        .team(savedTeam)
                        .playerNumber(1)
                        .displayName(row.getPlayer1Name().trim())
                        .status(PlayerStatus.INACTIVE)
                        .build();

                Player p2 = Player.builder()
                        .team(savedTeam)
                        .playerNumber(2)
                        .displayName(row.getPlayer2Name().trim())
                        .status(PlayerStatus.INACTIVE)
                        .build();

                playerRepository.save(p1);
                playerRepository.save(p2);

                gameStateService.initializeTeamGameState(savedTeam);

                teamsCreated++;
                playersCreated += 2;
                createdCodes.add(teamCode);

                log.info("Imported team '{}' (code: {}) with players '{}' and '{}'",
                        row.getTeamName(), teamCode, row.getPlayer1Name(), row.getPlayer2Name());

            } catch (Exception e) {
                importErrors.add("Row " + row.getRowNumber() + " ('" + row.getTeamName() + "'): " + e.getMessage());
                log.error("Failed to import team from row {}: {}", row.getRowNumber(), e.getMessage());
                throw new RuntimeException("Import failed at row " + row.getRowNumber() + ": " + e.getMessage(), e);
            }
        }

        int duplicatesSkipped = rows.size() - validRows.size();

        adminAuditService.logAction(
                principal,
                "EXCEL_TEAM_IMPORT",
                "Event #" + eventId,
                "Imported " + teamsCreated + " teams (" + playersCreated + " players) from Excel. "
                        + duplicatesSkipped + " rows skipped."
        );

        String summary = String.format(
                "Import complete: %d teams created, %d players created, %d rows skipped.",
                teamsCreated, playersCreated, duplicatesSkipped
        );

        return TeamImportResultDto.builder()
                .teamsCreated(teamsCreated)
                .playersCreated(playersCreated)
                .duplicatesSkipped(duplicatesSkipped)
                .errorsEncountered(importErrors.size())
                .createdTeamCodes(createdCodes)
                .errors(importErrors)
                .importTimestamp(Instant.now())
                .summary(summary)
                .build();
    }

    // ---- Internal Helpers ----

    private List<TeamImportRow> parseExcelFile(MultipartFile file) {
        List<TeamImportRow> rows = new ArrayList<>();

        try (InputStream is = file.getInputStream();
             Workbook workbook = new XSSFWorkbook(is)) {

            Sheet sheet = workbook.getSheetAt(0);
            if (sheet == null) {
                throw new IllegalArgumentException("Excel file contains no sheets.");
            }

            int lastRow = sheet.getLastRowNum();

            // Start from row 1 (row 0 is header)
            for (int i = 1; i <= lastRow; i++) {
                Row row = sheet.getRow(i);
                if (row == null) continue;

                String teamName = getCellStringValue(row.getCell(0));
                String player1Name = getCellStringValue(row.getCell(1));
                String player2Name = getCellStringValue(row.getCell(2));

                // Skip completely empty rows
                if (isBlank(teamName) && isBlank(player1Name) && isBlank(player2Name)) {
                    continue;
                }

                TeamImportRow importRow = TeamImportRow.builder()
                        .rowNumber(i + 1) // 1-indexed for user display
                        .teamName(teamName != null ? teamName.trim() : null)
                        .player1Name(player1Name != null ? player1Name.trim() : null)
                        .player2Name(player2Name != null ? player2Name.trim() : null)
                        .build();

                rows.add(importRow);
            }

        } catch (IllegalArgumentException e) {
            throw e;
        } catch (Exception e) {
            throw new IllegalArgumentException("Failed to parse Excel file: " + e.getMessage(), e);
        }

        return rows;
    }

    private void validateRows(List<TeamImportRow> rows, Long eventId) {
        Set<String> seenTeamNames = new HashSet<>();
        List<Team> existingTeams = teamRepository.findByEventId(eventId);
        Set<String> existingTeamNames = existingTeams.stream()
                .map(t -> t.getTeamName() != null ? t.getTeamName().toLowerCase().trim() : "")
                .collect(Collectors.toSet());

        for (TeamImportRow row : rows) {
            // Validate team name
            if (isBlank(row.getTeamName())) {
                row.addError("Team name is missing.");
            } else if (row.getTeamName().length() > MAX_TEAM_NAME_LENGTH) {
                row.addError("Team name exceeds " + MAX_TEAM_NAME_LENGTH + " characters.");
            } else if (containsInvalidCharacters(row.getTeamName())) {
                row.addError("Team name contains invalid characters.");
            } else {
                String normalizedName = row.getTeamName().toLowerCase().trim();
                if (seenTeamNames.contains(normalizedName)) {
                    row.addError("Duplicate team name within this import file.");
                } else if (existingTeamNames.contains(normalizedName)) {
                    row.addError("Duplicate: team name already exists in this event.");
                } else {
                    seenTeamNames.add(normalizedName);
                }
            }

            // Validate Player 1
            if (isBlank(row.getPlayer1Name())) {
                row.addError("Player 1 name is missing.");
            } else if (row.getPlayer1Name().length() > MAX_PLAYER_NAME_LENGTH) {
                row.addError("Player 1 name exceeds " + MAX_PLAYER_NAME_LENGTH + " characters.");
            }

            // Validate Player 2
            if (isBlank(row.getPlayer2Name())) {
                row.addError("Player 2 name is missing.");
            } else if (row.getPlayer2Name().length() > MAX_PLAYER_NAME_LENGTH) {
                row.addError("Player 2 name exceeds " + MAX_PLAYER_NAME_LENGTH + " characters.");
            }

            // Validate players are not the same
            if (!isBlank(row.getPlayer1Name()) && !isBlank(row.getPlayer2Name())
                    && row.getPlayer1Name().trim().equalsIgnoreCase(row.getPlayer2Name().trim())) {
                row.addError("Player 1 and Player 2 cannot have the same name.");
            }
        }
    }

    private void validateEventEditable(Event event) {
        if (event.getStatus() == EventStatus.RUNNING || event.getStatus() == EventStatus.COMPLETED) {
            throw new IllegalStateException(
                    "Cannot import teams: event is currently " + event.getStatus() + ".");
        }
    }

    private void validateFileFormat(MultipartFile file) {
        if (file == null || file.isEmpty()) {
            throw new IllegalArgumentException("No file uploaded.");
        }

        String filename = file.getOriginalFilename();
        if (filename == null || (!filename.endsWith(".xlsx") && !filename.endsWith(".xls"))) {
            throw new IllegalArgumentException(
                    "Invalid file format. Please upload an Excel file (.xlsx).");
        }

        // Max 5MB file size check
        if (file.getSize() > 5 * 1024 * 1024) {
            throw new IllegalArgumentException("File size exceeds 5MB limit.");
        }
    }

    private String getCellStringValue(Cell cell) {
        if (cell == null) return null;

        return switch (cell.getCellType()) {
            case STRING -> cell.getStringCellValue();
            case NUMERIC -> {
                double numVal = cell.getNumericCellValue();
                if (numVal == Math.floor(numVal) && !Double.isInfinite(numVal)) {
                    yield String.valueOf((long) numVal);
                }
                yield String.valueOf(numVal);
            }
            case BOOLEAN -> String.valueOf(cell.getBooleanCellValue());
            case FORMULA -> {
                try {
                    yield cell.getStringCellValue();
                } catch (Exception e) {
                    yield String.valueOf(cell.getNumericCellValue());
                }
            }
            default -> null;
        };
    }

    private boolean isBlank(String s) {
        return s == null || s.trim().isEmpty();
    }

    private boolean containsInvalidCharacters(String s) {
        // Allow alphanumeric, spaces, hyphens, underscores, dots, and common symbols
        return !s.matches("^[\\p{L}\\p{N}\\s\\-_.'&!@#]+$");
    }
}

File: backend/services/admin/teamExcelImportService.js
const xlsx = require('xlsx');
const eventRepository = require('../../repositories/eventRepository');
const teamRepository = require('../../repositories/teamRepository');
const playerRepository = require('../../repositories/playerRepository');
const gameStateService = require('../gameStateService');
const adminAuditService = require('./adminAuditService');
const { ResourceNotFoundException } = require('../../middleware/errorHandler');

const MAX_TEAM_NAME_LENGTH = 100;
const MAX_PLAYER_NAME_LENGTH = 100;

class TeamExcelImportService {
  async parseAndValidate(eventId, fileBuffer, originalFilename) {
    const event = await eventRepository.findById(eventId);
    if (!event) {
      throw new ResourceNotFoundException(`Event not found for ID ${eventId}`);
    }

    this._validateEventEditable(event);
    this._validateFileFormat(originalFilename, fileBuffer);

    const rows = this._parseExcelBuffer(fileBuffer);
    await this._validateRows(rows, eventId);

    const validCount = rows.filter(r => r.valid).length;
    const invalidCount = rows.filter(r => !r.valid).length;
    const duplicateCount = rows.filter(r => r.validationErrors.some(e => e.includes('Duplicate'))).length;

    const globalErrors = [];
    if (rows.length === 0) {
      globalErrors.push('No data rows found in the Excel file.');
    }
    if (validCount === 0 && rows.length > 0) {
      globalErrors.push('All rows have validation errors. No teams can be imported.');
    }

    const warnings = [];
    if (invalidCount > 0) {
      warnings.push(`${invalidCount} row(s) have validation errors and will be skipped during import.`);
    }

    return {
      totalRows: rows.length,
      validRows: validCount,
      invalidRows: invalidCount,
      duplicateRows: duplicateCount,
      rows,
      errors: globalErrors,
      warnings,
      importReady: validCount > 0 && globalErrors.length === 0
    };
  }

  async importTeams(eventId, fileBuffer, originalFilename, principal) {
    const event = await eventRepository.findById(eventId);
    if (!event) {
      throw new ResourceNotFoundException(`Event not found for ID ${eventId}`);
    }

    this._validateEventEditable(event);

    const rows = this._parseExcelBuffer(fileBuffer);
    await this._validateRows(rows, eventId);

    const validRows = rows.filter(r => r.valid);

    if (validRows.length === 0) {
      return {
        teamsCreated: 0,
        playersCreated: 0,
        duplicatesSkipped: rows.length,
        errorsEncountered: rows.length,
        errors: ['No valid rows to import.'],
        summary: 'Import aborted: no valid team rows found.'
      };
    }

    let teamsCreated = 0;
    let playersCreated = 0;
    const createdCodes = [];
    const importErrors = [];

    const currentTeamCount = await teamRepository.countByEventId(eventId);

    for (const row of validRows) {
      try {
        const countForCode = currentTeamCount + teamsCreated;
        const teamCode = await this._generateUniqueTeamCode(eventId, countForCode);

        const team = await teamRepository.createTeam({
          eventId: event.id,
          teamCode,
          teamName: row.teamName.trim(),
          status: 'REGISTERED',
          gameState: 'NOT_STARTED'
        });

        await playerRepository.create({
          teamId: team.id,
          playerNumber: 1,
          displayName: row.player1Name.trim(),
          status: 'INACTIVE'
        });

        await playerRepository.create({
          teamId: team.id,
          playerNumber: 2,
          displayName: row.player2Name.trim(),
          status: 'INACTIVE'
        });

        const fullTeam = await teamRepository.findById(team.id);
        await gameStateService.initializeTeamGameState(fullTeam);

        teamsCreated++;
        playersCreated += 2;
        createdCodes.push(teamCode);
      } catch (err) {
        importErrors.push(`Row ${row.rowNumber} ('${row.teamName}'): ${err.message}`);
        throw new Error(`Import failed at row ${row.rowNumber}: ${err.message}`);
      }
    }

    const duplicatesSkipped = rows.length - validRows.length;

    await adminAuditService.logAction(
      principal,
      'EXCEL_TEAM_IMPORT',
      `Event #${eventId}`,
      `Imported ${teamsCreated} teams (${playersCreated} players) from Excel. ${duplicatesSkipped} rows skipped.`
    );

    const summary = `Import complete: ${teamsCreated} teams created, ${playersCreated} players created, ${duplicatesSkipped} rows skipped.`;

    return {
      teamsCreated,
      playersCreated,
      duplicatesSkipped,
      errorsEncountered: importErrors.length,
      createdTeamCodes: createdCodes,
      errors: importErrors,
      importTimestamp: new Date().toISOString(),
      summary
    };
  }

  _parseExcelBuffer(buffer) {
    const workbook = xlsx.read(buffer, { type: 'buffer' });
    const firstSheetName = workbook.SheetNames[0];
    if (!firstSheetName) {
      const err = new Error('Excel file contains no sheets.');
      err.status = 400;
      throw err;
    }

    const sheet = workbook.Sheets[firstSheetName];
    const data = xlsx.utils.sheet_to_json(sheet, { header: 1, raw: false });

    const rows = [];
    // Row 0 is header, data starts at index 1
    for (let i = 1; i < data.length; i++) {
      const rowArr = data[i] || [];
      const teamName = rowArr[0] ? String(rowArr[0]).trim() : '';
      const player1Name = rowArr[1] ? String(rowArr[1]).trim() : '';
      const player2Name = rowArr[2] ? String(rowArr[2]).trim() : '';

      if (!teamName && !player1Name && !player2Name) {
        continue;
      }

      rows.push({
        rowNumber: i + 1,
        teamName: teamName || null,
        player1Name: player1Name || null,
        player2Name: player2Name || null,
        valid: true,
        validationErrors: []
      });
    }

    return rows;
  }

  async _validateRows(rows, eventId) {
    const seenTeamNames = new Set();
    const existingTeams = await teamRepository.findByEventId(eventId);
    const existingTeamNames = new Set(
      existingTeams.map(t => (t.teamName ? t.teamName.toLowerCase().trim() : ''))
    );

    for (const row of rows) {
      const addError = (msg) => {
        row.validationErrors.push(msg);
        row.valid = false;
      };

      if (!row.teamName || !row.teamName.trim()) {
        addError('Team name is missing.');
      } else if (row.teamName.length > MAX_TEAM_NAME_LENGTH) {
        addError(`Team name exceeds ${MAX_TEAM_NAME_LENGTH} characters.`);
      } else if (this._containsInvalidCharacters(row.teamName)) {
        addError('Team name contains invalid characters.');
      } else {
        const norm = row.teamName.toLowerCase().trim();
        if (seenTeamNames.has(norm)) {
          addError('Duplicate team name within this import file.');
        } else if (existingTeamNames.has(norm)) {
          addError('Duplicate: team name already exists in this event.');
        } else {
          seenTeamNames.add(norm);
        }
      }

      if (!row.player1Name || !row.player1Name.trim()) {
        addError('Player 1 name is missing.');
      } else if (row.player1Name.length > MAX_PLAYER_NAME_LENGTH) {
        addError(`Player 1 name exceeds ${MAX_PLAYER_NAME_LENGTH} characters.`);
      }

      if (!row.player2Name || !row.player2Name.trim()) {
        addError('Player 2 name is missing.');
      } else if (row.player2Name.length > MAX_PLAYER_NAME_LENGTH) {
        addError(`Player 2 name exceeds ${MAX_PLAYER_NAME_LENGTH} characters.`);
      }

      if (row.player1Name && row.player2Name && row.player1Name.trim().toLowerCase() === row.player2Name.trim().toLowerCase()) {
        addError('Player 1 and Player 2 cannot have the same name.');
      }
    }
  }

  _validateEventEditable(event) {
    if (event.status === 'RUNNING' || event.status === 'COMPLETED') {
      const err = new Error(`Cannot import teams: event is currently ${event.status}.`);
      err.status = 400;
      throw err;
    }
  }

  _validateFileFormat(filename, buffer) {
    if (!buffer || buffer.length === 0) {
      const err = new Error('No file uploaded.');
      err.status = 400;
      throw err;
    }
    if (filename && !filename.endsWith('.xlsx') && !filename.endsWith('.xls')) {
      const err = new Error('Invalid file format. Please upload an Excel file (.xlsx).');
      err.status = 400;
      throw err;
    }
    if (buffer.length > 5 * 1024 * 1024) {
      const err = new Error('File size exceeds 5MB limit.');
      err.status = 400;
      throw err;
    }
  }

  _containsInvalidCharacters(s) {
    return !/^[\p{L}\p{N}\s\-_.'&!@#]+$/u.test(s);
  }

  async _generateUniqueTeamCode(eventId, currentCount) {
    let index = currentCount + 1;
    while (true) {
      const code = `TEAM-${String(index).padStart(3, '0')}`;
      const existing = await teamRepository.findByEventIdAndTeamCode(eventId, code);
      if (!existing) {
        return code;
      }
      index++;
    }
  }
}

module.exports = new TeamExcelImportService();
